File: desktop/src/features/cos-follow-up/lib/cosFollowUp.ts
import type { RelayEvent } from "@/shared/api/types";
import {
  KIND_COS_FOLLOW_UP_COMMAND,
  KIND_COS_FOLLOW_UP_ITEM,
  KIND_COS_FOLLOW_UP_RECEIPT,
  KIND_DELETION,
} from "@/shared/constants/kinds";

export const COS_FOLLOW_UP_SCHEMA = "mac-workspace/cos-follow-up/v1";

export type CosFollowUpState = "needs-answer" | "ready-to-check" | "confirmed";

export type CosFollowUpAction =
  | "answer"
  | "confirm"
  | "reject"
  | "ready_to_check"
  | "reassign_confirmer";

export type CosFollowUpHumanAction = "answer" | "confirm" | "reject";

export type CosFollowUpPerson = {
  id: string | number | boolean;
  name: string;
};

export type CosFollowUpSource = {
  label: string;
  url: string;
};

export type CosFollowUpItem = {
  eventId: string;
  authorPubkey: string;
  channelId: string;
  assigneePubkey: string;
  id: string;
  jiraKey: string | null;
  title: string;
  question: string;
  evidence: string | null;
  state: CosFollowUpState;
  assignedPerson: CosFollowUpPerson;
  namedConfirmer: CosFollowUpPerson | null;
  version: number;
  permittedActions: CosFollowUpAction[];
  timestamps: Record<string, string | null>;
  deepLinks: {
    meetingFollowUp: string;
    jira: string | null;
    sources: CosFollowUpSource[];
  };
  createdAt: number;
};

export type CosFollowUpReceipt = {
  eventId: string;
  channelId: string;
  commandEventId: string;
  itemId: string;
  outcome: "accepted" | "rejected" | "conflict" | "failed";
  authoritativeVersion: number;
  message: string | null;
  code: string | null;
  retryable: boolean;
};

export type SeenActionableItem = {
  eventId: string;
  state: CosFollowUpState;
};

type JsonRecord = Record<string, unknown>;

const VALID_STATES = new Set<CosFollowUpState>([
  "needs-answer",
  "ready-to-check",
  "confirmed",
]);
const VALID_ACTIONS = new Set<CosFollowUpAction>([
  "answer",
  "confirm",
  "reject",
  "ready_to_check",
  "reassign_confirmer",
]);
const VALID_OUTCOMES = new Set<CosFollowUpReceipt["outcome"]>([
  "accepted",
  "rejected",
  "conflict",
  "failed",
]);

function record(value: unknown): JsonRecord {
  if (value === null || typeof value !== "object" || Array.isArray(value)) {
    throw new Error("Expected an object");
  }
  return value as JsonRecord;
}

function string(value: unknown, label: string): string {
  if (typeof value !== "string" || value.trim().length === 0) {
    throw new Error(`${label} must be a non-empty string`);
  }
  return value;
}

function nullableString(value: unknown): string | null {
  return typeof value === "string" && value.length > 0 ? value : null;
}

function safeAbsoluteUrl(value: unknown, label: string): string {
  const raw = string(value, label);
  let url: URL;
  try {
    url = new URL(raw);
  } catch {
    throw new Error(`${label} must be an absolute URL`);
  }
  const localHttp =
    url.protocol === "http:" &&
    ["localhost", "127.0.0.1", "[::1]"].includes(url.hostname);
  if (url.protocol !== "https:" && !localHttp) {
    throw new Error(`${label} must use HTTPS`);
  }
  return url.toString();
}

function nullableSafeAbsoluteUrl(value: unknown, label: string): string | null {
  return value === null || value === undefined || value === ""
    ? null
    : safeAbsoluteUrl(value, label);
}

function integer(value: unknown, label: string): number {
  if (!Number.isSafeInteger(value) || (value as number) < 0) {
    throw new Error(`${label} must be a non-negative safe integer`);
  }
  return value as number;
}

function person(value: unknown, label: string): CosFollowUpPerson {
  const raw = record(value);
  if (
    !["string", "number", "boolean"].includes(typeof raw.id) ||
    raw.id === null
  ) {
    throw new Error(`${label}.id must be a JSON scalar`);
  }
  return {
    id: raw.id as string | number | boolean,
    name: string(raw.name, `${label}.name`),
  };
}

function exactlyOneTag(event: RelayEvent, name: string): string {
  const values = event.tags
    .filter((tag) => tag[0] === name && typeof tag[1] === "string")
    .map((tag) => tag[1]);
  if (values.length !== 1) {
    throw new Error(`Follow-up event must contain exactly one ${name} tag`);
  }
  return values[0];
}

function parseContent(event: RelayEvent): JsonRecord {
  const raw = record(JSON.parse(event.content) as unknown);
  if (raw.schema !== COS_FOLLOW_UP_SCHEMA) {
    throw new Error("Unsupported COS follow-up schema");
  }
  return raw;
}

function parseActions(value: unknown): CosFollowUpAction[] {
  if (!Array.isArray(value)) {
    throw new Error("permitted_actions must be an array");
  }
  const actions = value.map((action) => {
    if (
      typeof action !== "string" ||
      !VALID_ACTIONS.has(action as CosFollowUpAction)
    ) {
      throw new Error("Unsupported COS follow-up action");
    }
    return action as CosFollowUpAction;
  });
  if (new Set(actions).size !== actions.length) {
    throw new Error("permitted_actions contains duplicates");
  }
  return actions;
}

function parseSources(value: unknown): CosFollowUpSource[] {
  if (value === undefined) return [];
  if (!Array.isArray(value)) throw new Error("sources must be an array");
  return value.map((source, index) => {
    const raw = record(source);
    return {
      label: string(raw.label, `sources[${index}].label`),
      url: safeAbsoluteUrl(raw.url, `sources[${index}].url`),
    };
  });
}

export function parseCosFollowUpItem(
  event: RelayEvent,
  expectedAssignee?: string,
): CosFollowUpItem {
  if (event.kind !== KIND_COS_FOLLOW_UP_ITEM) {
    throw new Error("Expected a COS follow-up item event");
  }
  const channelId = exactlyOneTag(event, "h");
  const itemId = exactlyOneTag(event, "d");
  const assigneePubkey = exactlyOneTag(event, "p").toLowerCase();
  if (expectedAssignee && assigneePubkey !== expectedAssignee.toLowerCase()) {
    throw new Error("COS follow-up item is assigned to a different identity");
  }
  const raw = parseContent(event);
  if (raw.id !== itemId) {
    throw new Error("COS follow-up item content id does not match d tag");
  }
  if (
    typeof raw.state !== "string" ||
    !VALID_STATES.has(raw.state as CosFollowUpState)
  ) {
    throw new Error("Unsupported COS follow-up state");
  }
  const prompt = record(raw.question_evidence);
  const links = record(raw.deep_links);
  const timestamps = record(raw.timestamps);

  return {
    eventId: event.id,
    authorPubkey: event.pubkey,
    channelId,
    assigneePubkey,
    id: itemId,
    jiraKey: nullableString(raw.jira_key),
    title: string(raw.title, "title"),
    question: string(prompt.question, "question_evidence.question"),
    evidence: nullableString(prompt.evidence),
    state: raw.state as CosFollowUpState,
    assignedPerson: person(raw.assigned_person, "assigned_person"),
    namedConfirmer:
      raw.named_confirmer === null || raw.named_confirmer === undefined
        ? null
        : person(raw.named_confirmer, "named_confirmer"),
    version: integer(raw.version, "version"),
    permittedActions: parseActions(raw.permitted_actions),
    timestamps: Object.fromEntries(
      Object.entries(timestamps).map(([key, value]) => [
        key,
        nullableString(value),
      ]),
    ),
    deepLinks: {
      meetingFollowUp: safeAbsoluteUrl(
        links.meeting_follow_up,
        "deep_links.meeting_follow_up",
      ),
      jira: nullableSafeAbsoluteUrl(links.jira, "deep_links.jira"),
      sources: parseSources(links.sources),
    },
    createdAt: event.created_at,
  };
}

export function projectLatestCosFollowUpItems(
  events: RelayEvent[],
  assigneePubkey: string,
  trustedBridgePubkey: string,
): CosFollowUpItem[] {
  const trustedAuthor = trustedBridgePubkey.toLowerCase();
  const latest = new Map<string, CosFollowUpItem>();
  for (const event of events) {
    if (event.pubkey.toLowerCase() !== trustedAuthor) continue;
    let item: CosFollowUpItem;
    try {
      item = parseCosFollowUpItem(event, assigneePubkey);
    } catch {
      continue;
    }
    latest.set(item.id, retainLatestCosFollowUpItem(latest.get(item.id), item));
  }
  return [...latest.values()].sort(
    (left, right) =>
      right.createdAt - left.createdAt || left.id.localeCompare(right.id),
  );
}

export function retainLatestCosFollowUpItem(
  current: CosFollowUpItem | undefined,
  candidate: CosFollowUpItem,
): CosFollowUpItem {
  if (
    !current ||
    candidate.version > current.version ||
    (candidate.version === current.version &&
      (candidate.createdAt > current.createdAt ||
        (candidate.createdAt === current.createdAt &&
          candidate.eventId.localeCompare(current.eventId) < 0)))
  ) {
    return candidate;
  }
  return current;
}

export function isCosFollowUpActionPermitted(
  item: CosFollowUpItem,
  action: CosFollowUpHumanAction,
): boolean {
  if (!item.permittedActions.includes(action)) return false;
  if (action === "answer") return item.state === "needs-answer";
  return item.state === "ready-to-check";
}

export function stateLabel(state: CosFollowUpState): string {
  if (state === "needs-answer") return "We need you";
  if (state === "ready-to-check") return "Does this look right?";
  return "Confirmed";
}

export function isNewlyActionableTransition(
  previous: SeenActionableItem | undefined,
  next: CosFollowUpItem,
): boolean {
  if (next.state === "confirmed") return false;
  if (previous?.eventId === next.eventId) return false;
  return previous?.state !== next.state;
}

export function parseCosFollowUpRemoval(event: RelayEvent): {
  channelId: string;
  itemId: string;
  targetEventId: string;
} {
  if (event.kind !== KIND_DELETION) {
    throw new Error("Expected a follow-up removal event");
  }
  return {
    channelId: exactlyOneTag(event, "h"),
    itemId: exactlyOneTag(event, "item"),
    targetEventId: exactlyOneTag(event, "e"),
  };
}

export function parseCosFollowUpReceipt(event: RelayEvent): CosFollowUpReceipt {
  if (event.kind !== KIND_COS_FOLLOW_UP_RECEIPT) {
    throw new Error("Expected a COS follow-up receipt");
  }
  const raw = parseContent(event);
  const outcome = exactlyOneTag(event, "outcome");
  if (!VALID_OUTCOMES.has(outcome as CosFollowUpReceipt["outcome"])) {
    throw new Error("Unsupported COS follow-up receipt outcome");
  }
  const retryable = raw.retryable === true;
  if (retryable && outcome !== "failed") {
    throw new Error("Only failed receipts can be retryable");
  }
  return {
    eventId: event.id,
    channelId: exactlyOneTag(event, "h"),
    commandEventId: exactlyOneTag(event, "e"),
    itemId: exactlyOneTag(event, "item"),
    outcome: outcome as CosFollowUpReceipt["outcome"],
    authoritativeVersion: integer(
      Number(exactlyOneTag(event, "version")),
      "version",
    ),
    message: nullableString(raw.message),
    code: nullableString(raw.code),
    retryable,
  };
}

export function buildCosFollowUpCommandInput({
  item,
  action,
  answer,
  comment,
}: {
  item: CosFollowUpItem;
  action: CosFollowUpHumanAction;
  answer?: string;
  comment?: string;
}): { kind: number; content: string; tags: string[][] } {
  if (!isCosFollowUpActionPermitted(item, action)) {
    throw new Error("That action is not permitted for the current item state");
  }
  if (action === "answer" && !answer?.trim()) {
    throw new Error("Please enter an answer");
  }
  return {
    kind: KIND_COS_FOLLOW_UP_COMMAND,
    tags: [
      ["h", item.channelId],
      ["item", item.id],
      ["action", action],
      ["expected-version", String(item.version)],
      ["e", item.eventId],
    ],
    content: JSON.stringify({
      schema: COS_FOLLOW_UP_SCHEMA,
      ...(answer?.trim() ? { answer: answer.trim() } : {}),
      ...(comment?.trim() ? { comment: comment.trim() } : {}),
    }),
  };
}

export function resolveAcceptedCosFollowUpProjection({
  receipt,
  action,
  itemId,
  items,
}: {
  receipt: CosFollowUpReceipt;
  action: CosFollowUpHumanAction;
  itemId: string;
  items: CosFollowUpItem[];
}):
  | { status: "pending" }
  | { status: "removed" }
  | { status: "updated"; item: CosFollowUpItem } {
  const current = items.find((item) => item.id === itemId);
  if (current && current.version >= receipt.authoritativeVersion) {
    return { status: "updated", item: current };
  }
  if (!current && (action === "answer" || action === "reject")) {
    return { status: "removed" };
  }
  return { status: "pending" };
}

File: mobile/lib/features/cos_follow_up/cos_follow_up.dart
import 'dart:convert';

import '../../shared/relay/relay.dart';

const cosFollowUpSchema = 'mac-workspace/cos-follow-up/v1';

enum CosFollowUpState {
  needsAnswer('needs-answer'),
  readyToCheck('ready-to-check'),
  confirmed('confirmed');

  final String wireValue;
  const CosFollowUpState(this.wireValue);

  static CosFollowUpState parse(Object? value) => values.firstWhere(
    (state) => state.wireValue == value,
    orElse: () => throw const FormatException('Unsupported follow-up state'),
  );
}

enum CosFollowUpHumanAction {
  answer('answer'),
  confirm('confirm'),
  reject('reject');

  final String wireValue;
  const CosFollowUpHumanAction(this.wireValue);
}

const _allActions = {
  'answer',
  'confirm',
  'reject',
  'ready_to_check',
  'reassign_confirmer',
};

Map<String, dynamic> _object(Object? value, String label) {
  if (value is! Map<String, dynamic>) {
    throw FormatException('$label must be an object');
  }
  return value;
}

String _string(Object? value, String label) {
  if (value is! String || value.trim().isEmpty) {
    throw FormatException('$label must be a non-empty string');
  }
  return value;
}

String? _nullableString(Object? value) =>
    value is String && value.isNotEmpty ? value : null;

String _safeAbsoluteUrl(Object? value, String label) {
  final raw = _string(value, label);
  final uri = Uri.tryParse(raw);
  final localHttp =
      uri?.scheme == 'http' &&
      const {'localhost', '127.0.0.1', '::1'}.contains(uri?.host);
  if (uri == null ||
      !uri.hasScheme ||
      !uri.hasAuthority ||
      (uri.scheme != 'https' && !localHttp)) {
    throw FormatException('$label must be an absolute HTTPS URL');
  }
  return uri.toString();
}

String? _nullableSafeAbsoluteUrl(Object? value, String label) =>
    value == null || value == '' ? null : _safeAbsoluteUrl(value, label);

int _integer(Object? value, String label) {
  if (value is! int || value < 0) {
    throw FormatException('$label must be a non-negative integer');
  }
  return value;
}

String _oneTag(NostrEvent event, String name) {
  final values = [
    for (final tag in event.tags)
      if (tag.length >= 2 && tag.first == name) tag[1],
  ];
  if (values.length != 1) {
    throw FormatException('Follow-up event needs exactly one $name tag');
  }
  return values.single;
}

Map<String, dynamic> _content(NostrEvent event) {
  final decoded = _object(jsonDecode(event.content), 'content');
  if (decoded['schema'] != cosFollowUpSchema) {
    throw const FormatException('Unsupported follow-up schema');
  }
  return decoded;
}

class CosFollowUpPerson {
  final Object id;
  final String name;

  const CosFollowUpPerson({required this.id, required this.name});

  factory CosFollowUpPerson.fromJson(Object? value, String label) {
    final json = _object(value, label);
    final id = json['id'];
    if (id is! String && id is! num && id is! bool) {
      throw FormatException('$label.id must be a JSON scalar');
    }
    return CosFollowUpPerson(
      id: id!,
      name: _string(json['name'], '$label.name'),
    );
  }
}

class CosFollowUpSource {
  final String label;
  final String url;

  const CosFollowUpSource({required this.label, required this.url});

  factory CosFollowUpSource.fromJson(Object? value, int index) {
    final json = _object(value, 'sources[$index]');
    return CosFollowUpSource(
      label: _string(json['label'], 'sources[$index].label'),
      url: _safeAbsoluteUrl(json['url'], 'sources[$index].url'),
    );
  }
}

class CosFollowUpDeepLinks {
  final String meetingFollowUp;
  final String? jira;
  final List<CosFollowUpSource> sources;

  const CosFollowUpDeepLinks({
    required this.meetingFollowUp,
    required this.jira,
    required this.sources,
  });
}

class CosFollowUpItem {
  final String eventId;
  final String authorPubkey;
  final String channelId;
  final String assigneePubkey;
  final String id;
  final String? jiraKey;
  final String title;
  final String question;
  final String? evidence;
  final CosFollowUpState state;
  final CosFollowUpPerson assignedPerson;
  final CosFollowUpPerson? namedConfirmer;
  final int version;
  final List<String> permittedActions;
  final Map<String, String?> timestamps;
  final CosFollowUpDeepLinks deepLinks;
  final int createdAt;

  const CosFollowUpItem({
    required this.eventId,
    required this.authorPubkey,
    required this.channelId,
    required this.assigneePubkey,
    required this.id,
    required this.jiraKey,
    required this.title,
    required this.question,
    required this.evidence,
    required this.state,
    required this.assignedPerson,
    required this.namedConfirmer,
    required this.version,
    required this.permittedActions,
    required this.timestamps,
    required this.deepLinks,
    required this.createdAt,
  });

  factory CosFollowUpItem.fromEvent(
    NostrEvent event, {
    String? expectedAssignee,
  }) {
    if (event.kind != EventKind.cosFollowUpItem) {
      throw const FormatException('Expected a follow-up item');
    }
    final channelId = _oneTag(event, 'h');
    final itemId = _oneTag(event, 'd');
    final assignee = _oneTag(event, 'p').toLowerCase();
    if (expectedAssignee != null &&
        assignee != expectedAssignee.toLowerCase()) {
      throw const FormatException('Item assigned to a different identity');
    }
    final json = _content(event);
    if (json['id'] != itemId) {
      throw const FormatException('Content id does not match d tag');
    }
    final prompt = _object(json['question_evidence'], 'question_evidence');
    final timestampJson = _object(json['timestamps'], 'timestamps');
    final linkJson = _object(json['deep_links'], 'deep_links');
    final rawActions = json['permitted_actions'];
    if (rawActions is! List) {
      throw const FormatException('permitted_actions must be an array');
    }
    final actions = rawActions.map((value) {
      if (value is! String || !_allActions.contains(value)) {
        throw const FormatException('Unsupported follow-up action');
      }
      return value;
    }).toList();
    if (actions.toSet().length != actions.length) {
      throw const FormatException('Duplicate follow-up action');
    }
    final rawSources = linkJson['sources'] ?? const [];
    if (rawSources is! List) {
      throw const FormatException('sources must be an array');
    }
    return CosFollowUpItem(
      eventId: event.id,
      authorPubkey: event.pubkey,
      channelId: channelId,
      assigneePubkey: assignee,
      id: itemId,
      jiraKey: _nullableString(json['jira_key']),
      title: _string(json['title'], 'title'),
      question: _string(prompt['question'], 'question_evidence.question'),
      evidence: _nullableString(prompt['evidence']),
      state: CosFollowUpState.parse(json['state']),
      assignedPerson: CosFollowUpPerson.fromJson(
        json['assigned_person'],
        'assigned_person',
      ),
      namedConfirmer: json['named_confirmer'] == null
          ? null
          : CosFollowUpPerson.fromJson(
              json['named_confirmer'],
              'named_confirmer',
            ),
      version: _integer(json['version'], 'version'),
      permittedActions: actions,
      timestamps: {
        for (final entry in timestampJson.entries)
          entry.key: _nullableString(entry.value),
      },
      deepLinks: CosFollowUpDeepLinks(
        meetingFollowUp: _safeAbsoluteUrl(
          linkJson['meeting_follow_up'],
          'deep_links.meeting_follow_up',
        ),
        jira: _nullableSafeAbsoluteUrl(linkJson['jira'], 'deep_links.jira'),
        sources: [
          for (var index = 0; index < rawSources.length; index++)
            CosFollowUpSource.fromJson(rawSources[index], index),
        ],
      ),
      createdAt: event.createdAt,
    );
  }

  bool isActionPermitted(CosFollowUpHumanAction action) {
    if (!permittedActions.contains(action.wireValue)) return false;
    if (action == CosFollowUpHumanAction.answer) {
      return state == CosFollowUpState.needsAnswer;
    }
    return state == CosFollowUpState.readyToCheck;
  }
}

List<CosFollowUpItem> projectLatestCosFollowUpItems(
  Iterable<NostrEvent> events,
  String assignee, {
  required String trustedBridgePubkey,
}) {
  final trustedAuthor = trustedBridgePubkey.toLowerCase();
  final latest = <String, CosFollowUpItem>{};
  for (final event in events) {
    if (event.pubkey.toLowerCase() != trustedAuthor) continue;
    try {
      final item = CosFollowUpItem.fromEvent(event, expectedAssignee: assignee);
      final current = latest[item.id];
      if (current == null ||
          item.version > current.version ||
          (item.version == current.version &&
              (item.createdAt > current.createdAt ||
                  (item.createdAt == current.createdAt &&
                      item.eventId.compareTo(current.eventId) < 0)))) {
        latest[item.id] = item;
      }
    } on FormatException {
      continue;
    }
  }
  final result = latest.values.toList()
    ..sort(
      (a, b) => b.createdAt.compareTo(a.createdAt) != 0
          ? b.createdAt.compareTo(a.createdAt)
          : a.id.compareTo(b.id),
    );
  return result;
}

class CosFollowUpCommand {
  final int kind;
  final String content;
  final List<List<String>> tags;

  const CosFollowUpCommand({
    required this.kind,
    required this.content,
    required this.tags,
  });
}

CosFollowUpCommand buildCosFollowUpCommand({
  required CosFollowUpItem item,
  required CosFollowUpHumanAction action,
  String? answer,
  String? comment,
}) {
  if (!item.isActionPermitted(action)) {
    throw StateError('Action not permitted for the current item state');
  }
  if (action == CosFollowUpHumanAction.answer &&
      (answer == null || answer.trim().isEmpty)) {
    throw ArgumentError('Please enter an answer');
  }
  return CosFollowUpCommand(
    kind: EventKind.cosFollowUpCommand,
    tags: [
      ['h', item.channelId],
      ['item', item.id],
      ['action', action.wireValue],
      ['expected-version', '${item.version}'],
      ['e', item.eventId],
    ],
    content: jsonEncode({
      'schema': cosFollowUpSchema,
      if (answer?.trim().isNotEmpty ?? false) 'answer': answer!.trim(),
      if (comment?.trim().isNotEmpty ?? false) 'comment': comment!.trim(),
    }),
  );
}

class CosFollowUpReceipt {
  final String eventId;
  final String channelId;
  final String commandEventId;
  final String itemId;
  final String outcome;
  final int authoritativeVersion;
  final String? message;
  final String? code;
  final bool retryable;

  const CosFollowUpReceipt({
    required this.eventId,
    required this.channelId,
    required this.commandEventId,
    required this.itemId,
    required this.outcome,
    required this.authoritativeVersion,
    required this.message,
    required this.code,
    required this.retryable,
  });

  factory CosFollowUpReceipt.fromEvent(NostrEvent event) {
    if (event.kind != EventKind.cosFollowUpReceipt) {
      throw const FormatException('Expected a follow-up receipt');
    }
    final outcome = _oneTag(event, 'outcome');
    if (!const {
      'accepted',
      'rejected',
      'conflict',
      'failed',
    }.contains(outcome)) {
      throw const FormatException('Unsupported receipt outcome');
    }
    final json = _content(event);
    final retryable = json['retryable'] == true;
    if (retryable && outcome != 'failed') {
      throw const FormatException('Only failed receipts can be retryable');
    }
    final version = int.tryParse(_oneTag(event, 'version'));
    if (version == null || version < 0) {
      throw const FormatException('Invalid receipt version');
    }
    return CosFollowUpReceipt(
      eventId: event.id,
      channelId: _oneTag(event, 'h'),
      commandEventId: _oneTag(event, 'e'),
      itemId: _oneTag(event, 'item'),
      outcome: outcome,
      authoritativeVersion: version,
      message: _nullableString(json['message']),
      code: _nullableString(json['code']),
      retryable: retryable,
    );
  }
}

class SeenCosFollowUp {
  final String eventId;
  final String state;

  const SeenCosFollowUp({required this.eventId, required this.state});
}

bool isNewlyActionableTransition(
  SeenCosFollowUp? previous,
  CosFollowUpItem next,
) {
  if (next.state == CosFollowUpState.confirmed) return false;
  if (previous?.eventId == next.eventId) return false;
  return previous?.state != next.state.wireValue;
}

String cosFollowUpStateLabel(CosFollowUpState state) => switch (state) {
  CosFollowUpState.needsAnswer => 'We need you',
  CosFollowUpState.readyToCheck => 'Does this look right?',
  CosFollowUpState.confirmed => 'Confirmed',
};
